FirebaseAdmin/FirebaseAdmin/Auth/UserProvider.cs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using Newtonsoft.Json;
+
+namespace FirebaseAdmin.Auth
+{
+    /// <summary>
+    /// Represents a user identity provider that can be associated with a Firebase user.
+    /// </summary>
+    public sealed class UserProvider
+    {
+        private string uid;
+
+        private string providerId;
+
+        /// <summary>
+        /// Gets or sets the user's unique ID assigned by the identity provider. This field is required.
+        /// </summary>
+        [JsonProperty("rawId")]
+        public string Uid
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(this.uid))
```

Comment:
Properties with built-in validation is always kindda awkward, and should be avoided whenever possible. Take this scenario for example:

```
var prov = new UserProvider();
if (prov.Uid == null)
{
  prov.Uid = someValue;
}
```

Most developers would expect the above to work, but it will throw because of our validation.

I suggest we introduce an internal nested class (similar to `UserImportRecordArgs.Request`) to handle the serialization. We can do all the validation during the type conversion.